### DISCUSSION TITLE
anansi-project/comicinfo

### DISCUSSION THREAD

**Discussion by SenorSmartyPants** [maforget/ComicRackCE#40] [General]
https://github.com/anansi-project/comicinfo

In case this project wasn't aware of the anansi project. They have been trying to document/advance the comicinfo.xml format. I'm not part of that project or sure of the buy in of other comic related projects but I wanted CRCE to be aware of it.

**Comment by maforget** [comment]
I've posted yesterday an updated schema based on the code. I've looked a little bit at that page trying to figure out the original schema. Like I said in that other post, maybe the dev had a schema, but ComicRack just exports it's class to an XML. The only thing you need to do to export a schema is just use this command `xsd.exe ComicRack.Engine.dll /t:ComicInfo` to export it from the source binaries. It comes with the .NET framework.

I've already added Tags to the ComicInfo earlier, other than that not really certain what is changed. The schema shows an enum for `ComicPageType`, while the existing one is just a string. I've looked at the code and it just outputs as a string, but still uses an enum. Doesn't change anything really, but to show in the exported schema I had to modify the code slightly so it shows in the export.

Other than that they also created an enum for Age Ratings instead of just a string, I don't know about this one, not all rating are the same in all the world.  But the have all of the Default ones.

All in all I don't see a lot of changes being made, the resulting files will be mostly the same. Only addition I see is that they added a Translator field, that could be a good idea to add to the program. There is also a GTIN, which I don't know what it is. Anything else?

- [x] Add Translator Field

[ComicInfo.xsd](https://github.com/maforget/ComicRackCE/files/14337764/ComicInfo.xsd.txt)

**Reply by DieselTech** [reply]
[GTIN ](https://en.wikipedia.org/wiki/Global_Trade_Item_Number ) is basically a UPC or EAN code. It should be a unique ID assigned to something globally that helps identify it. Amazon has their own version that people are used to, ASIN. 

As for adding stuff to the comicinfo.xml, I'm not going to say you should hold back in developing the app the way you want, but people are going to have some different `comicinfo.xml` files you might encounter after years and years of comicrack creating the "standard" and then abandoning it. For example, Kavita uses 2 non-standard fields that people have requested over time:

`<LocalizedSeries>` : Contains optional localized series name which will display in Kavita. Will allow searching for either series name. Will group files that have the localized name and the series name together as one series.

`<SeriesSort>` : The sort title to use for the Series. Kavita will prefer this over `<Series>`

Not only does Kavita support reading these fields, tools such as https://github.com/MangaManagerORG/Manga-Manager have been updated to write said fields. 

Where does this leave us? I don't know. I don't think it was ever really planned for ComicRack to make a comeback. People wrote it off as dead. But now that the CE project is going, we should at least be aware of all the different players in the space and how they work with the comicinfo.xml file.

**Reply by flaser** [reply]
I think the critical question is, does a xml with an _extended_ schema break parsing in either CR or Kavita?
Under _extended_, I mean it follows the old schema with some additional entries.
If not, then it could be OK to _add_ onto the schema _without_ breaking backwards compatibility.

This situation reminds me of how API Extensions are handled in a lot of projects.
Base API support is guaranteed, extensions are supported only on the specific variants (and guaranteed not to interfere with the Base).

Ideally the community should settle on one unified schema. However, if the current changes are not breaking base compatibility, support could be achieved in a gradual fashion.

~~For CR, I believe the main challenge is that as maforget wrote is the lack of an internal schema.~~
~~Pure serialization of the classes/objects strongly couples the generated XML's schema to the internal (in memory) representation.~~
~~This is not good as any changes for the sake of program improvement may break compatibility.~~

~~Going forward, I think CR may need a persistence layer that _translates_ to a schematic format.~~
~~This would decouple in-memory representation from the storage format.~~
~~It'd also be a chance to actually add formal, schematic verification.~~

UPDATE: I did some digging and it seems more straightforward now.

1. _ComicRackEngine\ComicInfo.cs_ holds the definition.
2. We can fine-tune serialization using attributes here:
    - https://learn.microsoft.com/en-us/dotnet/standard/serialization/attributes-that-control-xml-serialization
3. Finally, it should be possible to validate the generated XML against an external schema:
    - https://learn.microsoft.com/en-us/dotnet/standard/data/xml/schema-validation-using-xpathnavigator
    - Could we setup a test-case that'd let us know if changes to the CR class break compatibility?
    - (I'd need to dig into how schema validation happens to answer this definitely)

**Reply by maforget** [reply]
Pretty sure it will ignore the unknown tags. There weren't any problems with ComicRack as far as I know with other software. 

For now let's keep it as basic as it was. Although I don't know how other software handle the difference, I don't think there is any reason to be concerned here about serialization.

**Reply by SenorSmartyPants** [reply]
My concern would be that extended fields in comicinfo.xml would be ignored by CR on import, and then deleted when CR updates files.

**Reply by flaser** [reply]
I've been trying to track down whether this happens or not, based on the code.
For now, I'd tentatively say, yes this _can_ happen, as I've uncovered this call chain:

> If _updateComicFiles_ is enabled in Settings:
> ComicRack.Engine\QueueManager.cs:
> AddBookToFileUpdate-->WriteInfoToFileWithCacheUpdate-->WriteInfoToFile
> 
> ComicRack.Engine\ComicBook.cs:
> WriteInfoToFile-->StoreInfo
> 
> ComicRack.Engine\IO\Readers\ComicProvider.cs (base class w/o implemented method, always false):
> StoreInfo-->OnStoreInfo
> 
> ComicRack.Engine\IO\Readers\ArchiveComicProvider.cs (extending ComicProvider.cs, overriding method):
> OnStoreInfo-->WriteInfo
> 
> ComicRack.Engine\IO\Readers\Archive\SevenZipEngine.cs:
> WriteInfo-->UpdateComicInfo

Now this is where the "magic happens":
```
string text = Path.Combine(EngineConfiguration.Default.TempPath, Guid.NewGuid().ToString());
string text2 = Path.Combine(text, "ComicInfo.xml");
try
{
				Directory.CreateDirectory(text);
				using (Stream outStream = File.Create(text2))
				{
					comicInfo.Serialize(outStream);
				}
				string parameters2 = $"u -t{arg} \"{file}\" \"{text2}\"";
				if (ExecuteProcess.Execute(PackExe, parameters2, null, ExecuteProcess.Options.None).ExitCode == 0)
				{
					return true;
				}
}
```

A new file is created, ComicInfo is serialized into it, then 7zip is told to update the archive with this file.
**In other words, the old ComicInfo.xml will be overwritten.**

Now the question is this:
Are unknown XML nodes preserved in the in memory representation?

So far I see no sign of this happening.
**Therefore, my gut feeling is that unknown fields will be stripped from the XML.**

**Reply by maforget** [reply]
If that is all, then the fix is easy, already pushed. b255886d032c7ed12d1e74a56b96d2d1c8cd605e

Just followed this page, added a catch all for unknown elements that will be restored later. Tested with Export & Update, the unknown element stays.
https://learn.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlanyelementattribute?view=netframework-4.8

Edit: Moved the AllElement so that unknown tags are above the array of pages b7a19113663b87a727d079ae5fd5f6672f063b84

**Reply by flaser** [reply]
Awesome! Now we're assured that CR will play nice with other comic apps.

UPDATE:
Since we're doing a deep dive into XML serialization, I thought we could also look into **XML attributes**
> <Page **Image**="3" **ImageSize**="477854" **ImageWidth**="1360" **ImageHeight**="1920" />

Do we preserve unknown attributes too?

(I've been toying with the idea, to add a Chapter attribute to the Pages.
 This could provide tracking Chapters in TPBs or Manga tankoubons).

UPDATE:
Looks like we could use this Event to handle unknown attributes:
https://learn.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlserializer.unknownattribute?view=netframework-4.8#system-xml-serialization-xmlserializer-unknownattribute

> By default, after calling the [Deserialize](https://learn.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlserializer.deserialize?view=netframework-4.8) method, the [XmlSerializer](https://learn.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlserializer?view=netframework-4.8) ignores XML attributes of unknown types. However, you can use this event to handle such node types.
> 
> If the instance of the class being deserialized contains a field that returns an array of [XmlAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.xml.xmlattribute?view=netframework-4.8) objects and an [XmlAnyAttributeAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlanyattributeattribute?view=netframework-4.8) has been applied to the field, the [UnknownAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlserializer.unknownattribute?view=netframework-4.8#system-xml-serialization-xmlserializer-unknownattribute) event does not occur. Instead, all unknown XML attributes are collected into the array.

Seems like we need a way to handle unknown XML attributes for XML elements that _do_ have a mapped member e.g. like my proposed "Chapter" entry for Pages.
Using an event-handler, we could collate the unknown attributes into a class dictionary.
Then, during serialization, do a lookup for each element to reinsert these.

**Reply by maforget** [reply]
I can always add an `XmlAnyAttribute` to the  `ComicPageInfo`. Although I don't know if any unknown tags exists. We already have the Bookmark that does basically the same as a chapter in ComicRack.

**Reply by flaser** [reply]
I'm not sure as according its desc _XmlAnyAttribute_ only captures whole XML Elements without mapping, not XML attributes.
(The name is a bit misleading, as I think the Attribute here refers to the C# concept instead the XML one >_<)

What I'm talking about is an unknown XML attribute (e.g. ImageWidth) in a parsed XML Element (e.g. Page).
If I'm correct, any unknown XML attribute of parsed XML elements should be quietly dropped at the moment.

UPDATE 1:
- Did a quick test, **unknown XML attributes in parsed XML elements are stripped out.**

UPDATE 2:
- ~I've seen parsing breaking and comicinfo reset if I add unknown XML elements. I'll do more investigation.~
- This was caused by malformed XML, see UPDATE 5 below.

UPDATE 3:
- Recognized source of some of the weirdness I'm seeing.
- The 1st time a file is loaded is when it's stored in memory.
- Unless I force CR to reload from file (e.g. restart app), it will keep & serialize this state.

UPDATE 4:
- I can confirm multiple unknown XML Elements are stored & reserialized.

| Type              | Attributes | Example                                                                                                           |
| -------------- | ----------- | ---------------------------------------------------------------------------------- |
| regular          | no             | `<MyElement1>LoremIpsum</MyElement1>`                                               |
| regular          | yes            | `<MyElement2 Attr="x">LoremIpsum</MyElement1>`                                  |
| self-closing   | no             |  `<MyElement3 />`                                                                                           |
| self-closing   | yes            |  `<MyElement4 Height="1" Weight="2"/>`                                                     |
| nested          | no             |   `<MyElement5><Nest1>Lorem Ipsum</Nest1></MyElement5>`               |
| nested          | yes            |   `<MyElement5><Nest2 Attr="x">Lorem Ipsum</Nest2></MyElement5>` |

- I couldn't reproduce breaking of parsing by adding extra XML attributes to existing XML elements.
- I've rechecked and can confirm that these extra XML attributes get stripped out.

UPDATE 5:
- A malformed XML will break parsing --> **This** was causing my earlier issue
- [BUG!](https://github.com/maforget/ComicRackCE/issues/45#issue-2147344639) CR will always update a malformed ComicInfo.xml<br>(Even if the user choosed not update files outside the DB)

**Reply by maforget** [reply]
I think you are confused, I added a `XmlAnyElementAttribute` to the `ComicInfo` class. The [`XmlAnyAttributeAttribute`](https://learn.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlanyattributeattribute?view=netframework-4.8) is the same but for attributes, since it hasn't been added then it's normal that they are ignored. You just don't write the last Attribute when you use it. 

No need to use an Event and then pick apart the Unknown, the idea is to just have someplace to store them. I was talking adding it to the `ComicPageInfo` because it's the only place that actually uses attributes, the `ComicInfo` only uses Elements. 

The idea is to make sure CR doesn't remove some of the other program tags, not have some Frankenstein of possible XML format. The `XmlElement` that stores the Unparsed can contain attribute, that is why you are seeing the Attributes being save for Unknown Tags, but not the rest. Normally Attribute would be Properties of a subclass, you wouldn't have say a `Title` Element have an attribute, because it's supposed to be a string. That is why in the schema only The `ComicPageInfo` uses attributes. 

Of course in a normal XML you could have a string have an attribute, but this isn't an XML => OBJECT it's an OBJECT => XML. It's a representation of the object in memory and has such, some of these theoretical wouldn't happen. To do what you want would require use to actually inspect each `XmlElement` & `XmlNode`, figure out if the element has a corresponding property. Then what if it has other unexpected elements after them, figure out a way to store them so they can be restored in the correct order. This is a mess and it isn't really required because the XmlSerializer does it for us and we shouldn't encounter these anyway. I don't know how the other software do it, but it's probably something similar to serializing.

Now I will have to check on that bug, because if you don't update the file it shouldn't touch the file regardless of it being malformed. But then again why would it be malformed, if it created by the serializer and is just a representation of an object?

**Reply by flaser** [reply]
You're correct, I was mixing things up with [XmlAnyElementAtttribute](https://learn.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlanyelementattribute?view=netframework-4.8).
Indeed, [XmlAnyAttributeAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlanyattributeattribute?view=netframework-4.8) should do exactly what is required.
(The example even demonstartes this very use - d'oh, better look more carefully next time!)

You're also correct there's no need to support any arbitrary XML==>Object schema.
I was indeed thinking of ways how to preserve data added by other programs.
As for how it can get malformed? Simplest is user error, or a bug in an external scraper.

**Reply by maforget** [reply]
The change also as the effect of storing the unparsed elements in the database.

**Reply by maforget** [reply]
Added the Translator tag

**Comment by therobbiedavis** [comment]
The point of the Anansi Project is to **try** and move the comicinfo schema forward as a standard for the community at large. The issue is it’s been multiple years without much or any progress forward and the current schema version is sorely lacking in some areas (hence Kavita moving forward with supporting non-standard fields).

I don’t think the Anansi Project as it stands right now is the way forward, but I think the idea behind it is. The culmination of all the key players working towards a better standardized schema for all.

**Comment by maforget** [comment]
A User on Reddit suggested adding the reading progress to the xml. There is the Current Page/Last Page Read properties that are related. I don't have any opinion on that either way. There was a reason it wasn't added, especially if sharing files. Good Idea? Bad Idea?

**Reply by DieselTech** [reply]
I wouldn't mind if comicrack kept track of that field on it's own for users files, but I don't think writing it into the comicinfo.xml is the way to go. As that is meant to be data about the issue itself, not necessarily user specific data.

I just see a lot of confusion around people going to open a file and it not starting at page one because the .xml has a progress stored in it.

**Reply by flaser** [reply]
What about creating a separate XML file inside the archive?
This could be ReadingStatus.xml or something similar with a class equivalent to serialize the same way ComicInfo is treated.
We could collected related primitive type fields from the ComicBook class into the new ReadingStatus class.
The ComicBook class in turn can just hold an instance of this ReadingStatus class as a member.

**Reply by Totengeist** [reply]
What is the benefit to using a different file in the same archive?

**Reply by maforget** [reply]
I personally have a similar thinking as @DieselTech, that it doesn't relate to the file itself. Although @flaser is an idea, it would still be remembered when opening the file, so you would have an annoying popup asking to resume the last position or a setting that 90% of users won't know about. 

Remember Users are dump (sorry but a large majority). I had to help a lot of people on Reddit that kept saying that ComicRack didn't remember their reading progress, just because they had disabled the option to track current page status. Like a teacher I had once said, just make sure that things are idiot proof.

That being said, calibre does save the reading position in the book itself, it saves a calibre_bookmark that contains some text that is base64 encoded with the location & date. And I don't know if I never stumbled upon a file that already had a bookmark set.

**Reply by DieselTech** [reply]
I feel like for something like this it goes into the re-write direction that we commented on in another thread. If CR's own sqlite database held that info then it would be store data like that. It would then also be easier for users to move in or out of different systems, without worrying if the reader they are using supports the very specific tag. Scripts could be made to query the db directly so that progress from other tracking apps like Kavita / Komga / Stump / Codex can be either inserted or exported.

**Reply by maforget** [reply]
> I feel like for something like this it goes into the re-write direction that we commented on in another thread. If CR's own sqlite database held that info then it would be store data like that. It would then also be easier for users to move in or out of different systems, without worrying if the reader they are using supports the very specific tag. Scripts could be made to query the db directly so that progress from other tracking apps like Kavita / Komga / Stump / Codex can be either inserted or exported.

This is irrelevant to either the form of data storage. The info is already stored in the db, that it is in sqlite or not is irrelevant. The question is do we make that info also saved in the file? That file is only used when restoring or sharing a comic. It is never used by ComicRack directly beside storing them.

Plugins just refer to the data that the `ComicBook` object contains. When CR loads a file the source of that data is usually the database, if it's not in the database, it will check if a XML exists inside, the NTFS alternate stream or either a side XML beside the file. The source of the data is irrelevant and should be transparent to the program. Script can already access all the info in the `ComicBook`, regardless of the source. You have `ComicBook` that inherits `ComicInfo`, the info that are in `ComicInfo` are saved in `ComicInfo.xml` (hence the name). While the data in `ComicBook` is just saved in the DB (which includes `ComicInfo`). Having the data saved in the XML is only a question of transferring the property from one class to the other.

My question is not whether it is compatible with other software, with all the different version I don't believe it is a problem. Normal XML already ignore unknown tags and to be frank it is these program problem. We can play fair with other and try not to erase tags we don't know like I did already, Maybe because I don't use these software, but my focus is just ComicRack. If I can do anything to fix a compatibility problem and it isn't a big deal, I will. But I ain't gonna bother what might happen if I implement what is best for ComicRack.

As for what another file would do, it might help prevent any compatibility problem with the OG ComicRack or other software, if there is any reason for concern. It might be easy to decouple the reading status from the `ComicBook` object, but since we would need to rewire the program to access this field anyway, so it won't change nothing.

**Reply by azuravian** [reply]
What is the purpose of writing the reading progress to the file itself instead of just keeping it in the db?

**Reply by maforget** [reply]
> What is the purpose of writing the reading progress to the file itself instead of just keeping it in the db?

That user reason was that he frequently reinstall his windows (and doesn't seem to backup the db) and would like reading progress to be saved. 

Although not the best reason I could see the usage, but like others have said I don't know if it's place is in the file itself.

**Reply by SenorSmartyPants** [reply]
Reading progress is about the user, not the book. It makes no sense to write it to the book if you consider multiple users.

**Reply by Nadiar** [reply]
Would make more sense to have a separate file in the directory with the cbz not inside of it, and probably as a plugin.  You could just do a <filename-username.json> (format doesn't matter, just for illustration I'm assuming Jason) document to store "local" information.

**Comment by TheRealC0unt** [comment]
It might be useful to get in touch with the author of Komga server to coordinate about comicinfo tags.
At least he as well participates in the Anansi Project, if I'm not mistaken, and Komga is pretty widespread these days.
I am used to tag all my Comics and mangas with CR prior to out them on my Komga server, for example.

**Reply by maforget** [reply]
All we have done is add Tags to comicinfo which the Anansi project also does and support for any unknown tags that could be there. I don't know what else there is to add or change to it. Not planning on doing any adding any new tags to it. CR => Komga will work correctly like it does. Any other Software => CR should keep these data also.

**Comment by killo3967** [comment]
Here is a discussion about GTIN in the Anasi Project.

https://github.com/anansi-project/comicinfo/issues/50

I think it's an important filed. Basically it's the id's of each web that contains metadata that you scrape.

**Reply by maforget** [reply]
Right now the 2.1 draft shows that GTIN is a single string entry and the proposal would be to have multiples identifiers and add an attribute called type where we can set which type of identifier it is? 

Has anyone actually did changes in that regard or to even support the GTIN single entry?

No point to implement a new field when there isn't anything certain yet.

**Reply by killo3967** [reply]
Your right, so in add into the anasi-project DTIN an entry to proposing that GTIN has multiple tipes of entries as Calibre do with the ID's of the books.

**Comment by maforget** [comment]
I've added support for reading the `MetronInfo.xml` from the [MetronInfo](https://github.com/Metron-Project/metroninfo) project. Doing so I've made change for support for multiple xml file. So it would be possible to have multiple types of files with various schema be supported.

The internal of the program are still the ComicInfo, so for now this only maps the corresponding fields from the `MetronInfo.xml` to the `ComicInfo.xml`. There is an order specified in the code, so it will prefer the `ComicInfo.xml` if both are present and really only read `MetronInfo.xml` when it is the only file present. Also a note that exporting will delete any `MetronInfo.xml` file but keep the metadata inside the `ComicInfo.xml`.
